llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -8144,9 +8147,8 @@ void BoUpSLP::buildTree_rec(ArrayRef<Value *> VL, unsigned Depth,
   }
 
   // Don't handle scalable vectors
-  if (S.getOpcode() == Instruction::ExtractElement &&
-      isa<ScalableVectorType>(
-          cast<ExtractElementInst>(S.getMainOp())->getVectorOperandType())) {
+  if (auto *EE = dyn_cast_if_present<ExtractElementInst>(S.getMainOp());
```

Comment:
```suggestion
  if (auto *EE = dyn_cast<ExtractElementInst>(S.getMainOp());
```

Reply:
Cannot. S may not be valid.

Reply:
Then you cannot use getMainOp

Reply:
1. The origin code check MainOp opcode (`S.getOpcode() == Instruction::ExtractElement`) and use `cast` later. I think use `dyn_cast_if_present` is better.
```
  if (S.getOpcode() == Instruction::ExtractElement &&
      isa<ScalableVectorType>(
          cast<ExtractElementInst>(S.getMainOp())->getVectorOperandType())) {
```
2. There is a case which uses `isa_and_present`. Follow the same idea, I use `dyn_cast_if_present` here.
```
  bool AreAllSameInsts = AreAllSameBlock || AreScatterAllGEPSameBlock;
  if (!AreAllSameInsts || (!S && allConstant(VL)) || isSplat(VL) ||
      (isa_and_present<InsertElementInst, ExtractValueInst, ExtractElementInst>(
           S.getMainOp()) &&
       !all_of(VL, isVectorLikeInstWithConstOps)) ||
      NotProfitableForVectorization(VL)) {
```

Reply:
1. I.e. the original code check if S is valid, has the opcode ExtractElement and then checks the vector type. getMainOp will crash, if S is invalid. If S is valid, getMainOp cannot return nullptr. dyn_cast_if_present should be used here, just a dyn_cast should be enough. getMainOp just cannot return nullptr.
2. It should be fixed, need to replace by `S and isa<....>(S.getMainOp())`. It worked before, currently it will crash, if S is invalid

Reply:
Currently some code use isa some code use getOpcode. I think we should either

- use isa, dyn_cast, isa_and_present and dyn_cast_if_present for getMainOp
- use getOpcode and check whether the opcode is what we want

to make the code consistency.
However, isa (and others) can check `isa_and_present<InsertElementInst, ExtractValueInst, ExtractElementInst>` with less code. Use `getOpcode` will write a lot of code in this case.

Reply:
It won't work, it will crash the compiler, `_and_present` are not required

Reply:
It can work. See https://github.com/llvm/llvm-project/pull/120210/commits/c9ccb5577ac8c9994199bd761231ab1b1fd41321.
`isa_and_present<>` = `S.valid() && isa<>`

Reply:
It should not work. `getMainOp()` and `getAltOp()` should crash if requested from invalid state

Reply:
OK. See https://github.com/llvm/llvm-project/pull/120217/commits/1c16e5cc54bcbd24eb51f2609f3a098b27d80b96